rclcpp/src/rclcpp/parameter_events_subscriber.cpp

```diff
@@ -0,0 +1,160 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "rclcpp/parameter_events_subscriber.hpp"
+
+namespace rclcpp
+{
+
+ParameterEventsSubscriber::ParameterEventsSubscriber(
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
+  rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics,
+  rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
+  const rclcpp::QoS & qos)
+: node_base_(node_base),
+  node_topics_(node_topics),
+  node_logging_(node_logging),
+  qos_(qos)
+{}
+
+void ParameterEventsSubscriber::add_namespace_event_subscriber(const std::string & node_namespace)
+{
+  if (std::find(node_namespaces_.begin(), node_namespaces_.end(),
+    node_namespace) == node_namespaces_.end())
+  {
+    node_namespaces_.push_back(node_namespace);
+    auto topic = join_path(node_namespace, "parameter_events");
+    RCLCPP_DEBUG(node_logging_->get_logger(), "Subscribing to topic: %s", topic.c_str());
+
+    auto event_sub = rclcpp::create_subscription<rcl_interfaces::msg::ParameterEvent>(
+      node_topics_, topic, qos_,
+      std::bind(&ParameterEventsSubscriber::event_callback, this, std::placeholders::_1));
+
+    event_subscriptions_.push_back(event_sub);
+  }
+}
+
+void ParameterEventsSubscriber::set_event_callback(
+  std::function<void(const rcl_interfaces::msg::ParameterEvent::SharedPtr &)> callback,
+  const std::string & node_namespace)
+{
+  std::string full_namespace;
+  if (node_namespace == "") {
+    full_namespace = node_base_->get_namespace();
+  } else {
+    full_namespace = resolve_path(full_namespace);
+  }
+  add_namespace_event_subscriber(full_namespace);
+  user_callback_ = callback;
+}
+
+void ParameterEventsSubscriber::register_parameter_callback(
+  const std::string & parameter_name,
+  std::function<void(const rclcpp::Parameter &)> callback,
+  const std::string & node_name)
+{
+  auto full_node_name = resolve_path(node_name);
+  add_namespace_event_subscriber(split_path(full_node_name).first);
+  parameter_callbacks_[{parameter_name, full_node_name}] = callback;
+}
+
+bool ParameterEventsSubscriber::get_parameter_from_event(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event,
+  rclcpp::Parameter & parameter,
+  const std::string parameter_name,
+  const std::string node_name)
+{
+  if (event->node == resolve_path(node_name)) {
+    rclcpp::ParameterEventsFilter filter(event, {parameter_name},
+      {rclcpp::ParameterEventsFilter::EventType::NEW,
+        rclcpp::ParameterEventsFilter::EventType::CHANGED});
+    if (!filter.get_events().empty()) {
+      const auto & events = filter.get_events();
+      auto param_msg = events[events.size() - 1].second;
```

Comment:
nit: I forgot in the original review, but `std::vector` has a [back](https://en.cppreference.com/w/cpp/container/vector/back) method.